File: golang/binary-tree-general/flatten_binary_tree.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func flatten(root *TreeNode) {
	if root == nil {
		return
	}

	current := root

	for current != nil {
		if current.Left != nil {
			right_most := current.Left

			for right_most.Right != nil {
				right_most = right_most.Right
			}

			right_most.Right = current.Right

			current.Right = current.Left

			current.Left = nil
		}

		current = current.Right
	}
}

File: python/binary-tree-general/flatten_binary_tree.py
from typing import Optional


class TreeNode:
    def __init__(self, val: int = 0, left: None = None, right: None = None) -> None:
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def flatten(self, root: Optional[TreeNode]) -> None:
        if not root:
            return None

        current = root

        while current:
            if current.left:
                right_most = current.left

                while right_most.right:
                    right_most = right_most.right

                right_most.right = current.right

                current.right =  current.left

                current.left = None

            current = current.right
